File: backend/src/main/java/com/github/jadamon42/family/model/Partnership.java
package com.github.jadamon42.family.model;

import com.github.jadamon42.family.util.UUIDStringConverter;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Value;
import lombok.With;
import org.springframework.data.neo4j.core.convert.ConvertWith;
import org.springframework.data.neo4j.core.schema.GeneratedValue;
import org.springframework.data.neo4j.core.schema.Id;
import org.springframework.data.neo4j.core.schema.Node;
import org.springframework.data.neo4j.core.schema.Relationship;
import org.springframework.data.neo4j.core.support.UUIDStringGenerator;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

import static com.github.jadamon42.family.util.PatchHelper.set;

@Value
@With
@Builder
@AllArgsConstructor
@Node("Partnership")
public class Partnership {
    @Id @GeneratedValue(generatorClass = UUIDStringGenerator.class)
    @ConvertWith(converter = UUIDStringConverter.class)
    UUID id;
    String type;
    LocalDate startDate;
    LocalDate endDate;

    @Relationship(value = "PARTNER_IN", direction = Relationship.Direction.INCOMING)
    @Builder.Default
    List<Person> partners = List.of();
    @Relationship(value = "BEGAT", direction = Relationship.Direction.OUTGOING)
    @Builder.Default
    List<Person> children = List.of();

    public static Partnership fromRequest(PartnershipRequest request) {
        Partnership.PartnershipBuilder builder = Partnership.builder();
        set(builder, PartnershipBuilder::type, request.getType());
        set(builder, PartnershipBuilder::startDate, request.getStartDate());
        set(builder, PartnershipBuilder::endDate, request.getEndDate());
        return builder.build();
    }
}

File: frontend/src/renderer/models/Partnership.ts
import type Person from './Person';

class Partnership {
  id: string;
  type: string;
  startDate: Date;
  endDate: Date;
  partners: Person[];
  children: Person[];

  constructor(
    id: string,
    type: string,
    startDate: Date,
    endDate: Date,
    partners: Person[],
    children: Person[]
  ) {
    this.id = id;
    this.type = type;
    this.startDate = startDate;
    this.endDate = endDate;
    this.partners = partners;
    this.children = children;
  }

  getPartnershipData() {
    return {
      id: this.id,
      type: this.type,
      startDate: this.startDate,
      endDate: this.endDate,
      partnerIds: this.partners.map((partner: Person) => partner.id),
      childrenIds: this.children.map((child: Person) => child.id),
    };
  }
}

export default Partnership;
